src/brick_game/snake/snake.cpp

```diff
@@ -0,0 +1,238 @@
+    switch (direction_) {
+      case UP:
+        --tmp.y;
+        break;
+      case DOWN:
+        ++tmp.y;
+        break;
+      case LEFT:
+        --tmp.x;
+        break;
+      case RIGHT:
+        ++tmp.x;
+        break;
+    }
+    process_move(tmp);
+    clock1_ = clock2_;
+  }
+}
+
+void s21::Snake::process_move(Position new_head) {
+  if (new_head == apple_) {
+    eatApple();
+  } else {
+    field_[snake_.back().y][snake_.back().x] = 0;
+    snake_.pop_back();
+  }
+  if (field_[new_head.y][new_head.x] == body || new_head.x == -1 ||
+      new_head.x == FIELD_WIDTH || new_head.y == FIELD_HEIGHT ||
+      new_head.y == -1) {
+    state_ = GAMEOVER;
+  } else {
+    field_[snake_.front().y][snake_.front().x] = body;
+    snake_.push_front(new_head);
+    field_[snake_.front().y][snake_.front().x] = head;
+  }
+}
+
+void s21::Snake::eatApple() {
+  ++game_.score;
+  if (game_.score > game_.high_score) {
+    updateHighscore();
+  }
+  if (game_.score <= MAX_SCORE_TO_LVLUP) game_.speed = game_.level = 1 + (game_.score) / POINTS_TO_LVLUP;
+  if (game_.score == MAX_SCORE) {
+    state_ = GAMEOVER;
+  } else
+    state_ = SPAWN;
+}
+
+void s21::Snake::spawnApple() {
+  std::vector<Position> vec;
+  for (int i = 0; i < FIELD_HEIGHT; ++i) {
+    for (int j = 0; j < FIELD_WIDTH; ++j) {
+      if (field_[i][j] == 0) {
+        vec.push_back({i, j});
+      }
+    }
+  }
+  apple_ = {vec[rand() % vec.size()]};
+
+  field_[apple_.y][apple_.x] = apple;
+  state_ = MOVE;
+}
+
+void s21::Snake::checkTick() {
+  time_to_fall_ = 0;
+  clock2_ = clock();
+
+  if ((double)(clock2_ - clock1_) / CLOCKS_PER_SEC >=
+      move_timer[game_.speed - 1]) {
+    time_to_fall_ = 1;
+  }
+}
+
+void s21::Snake::updateHighscore() {
+  game_.high_score = game_.score;
+  FILE *file = fopen(SNAKE_HIGHSCORE, "rb+");
+  if (file) {
+    char score[12];
+    if (fgets(score, sizeof(score), file)) {
+    }
+    int highscore = atoi(score);
+    if (game_.high_score > highscore) {
+      snprintf(score, sizeof(score), "%d", game_.high_score);
+      fseek(file, 0, SEEK_SET);
+      if (ftruncate(fileno(file), 0) == 0) {
+        fputs(score, file);
+      }
+    }
+    fclose(file);
+  }
+}
+
+void s21::Snake::allocateGameInfo() {
+  game_.next = nullptr;
+  game_.score = 0;
+  game_.high_score = 0;
+  game_.level = 1;
+  game_.speed = 1;
+  game_.pause = START;
+  game_.field = new int *[20];
+  for (int i = 0; i < FIELD_HEIGHT; i++) {
+    game_.field[i] = new int[10];
+    for (int j = 0; j < FIELD_WIDTH; j++) {
+      game_.field[i][j] = 0;
+    }
+  }
+  FILE *file = fopen(SNAKE_HIGHSCORE, "rb+");
+  if (file != nullptr) {
+    char score[FIELD_WIDTH] = {'\0'};
+    if (fgets(score, sizeof(score), file)) game_.high_score = atoi(score);
+    fclose(file);
+  } else {
+    file = fopen(SNAKE_HIGHSCORE, "wb+");
+    if (file) {
+      fputs("0", file);
+      fflush(file);
+    }
+  }
+}
+
+void s21::Snake::freeGameInfo() {
+  if (game_.field) {
```

Comment:
В плюсовом коде можно и даже нужно использовать "множественные ретёрны".

```suggestion
  if (!game_.field) return;
```

Подобный код позволит избежать излишнюю вложенность и ветвления. Упростит код.

Reply:
вау